[build] Add `symlink_system_library` calls for CSL

staticfloat opened the pull request:
We handle `USE_SYSTEM_CSL=1` the same as `USE_BINARYBUILDER_CSL=0`, so
we should always install the libraries locally.  Arguably we could
symlink instead of copying, but the benefit of isolation where the
user can then run `make binary-dist` is still evident.

Fixes #40201

staticfloat commented:
Fixes https://github.com/JuliaLang/julia/issues/40201

yuyichao reviewed (changes requested):
System libraries must not be copied and must be symlinked instead.
If you want to support creating a portable build out of system library, that should be handled with a different option.

vtjnash commented:
Previously, we've been handling these cases in base/Makefile. Perhaps not really the most consistent place to handle it, since many of those are actually present for `_jll` files, but we should at least keep the handling consistent and together, where ever that may be.

staticfloat committed:
[build] Add `symlink_system_library` calls for CSL

We differentiate between a `USE_BINARYBUILDER_CSL=0` configuration
(where we copy the CSL libraries into the Julia build tree) versus a
`USE_SYSTEM_CSL=1` configuration (where we symlink them in a manner
similar to other system libararies)

staticfloat commented:
Alright, new implementation that is a little bit more work (mostly because we can't find `libgcc_s` without adding the SOVERSION to the end, which required re-architecting some Makefile functions) but what I have now gives us three separate configurations:

1) `USE_BINARYBUILDER_CSL=1`: behavior unchanged, downloads and bundles the CSL from Yggdrasil.  This is the default.

2) `USE_BINARYBUILDER_CSL=0`: Copies the libraries from the current compiler suite into `usr/lib`, as this is the closest we can get to a "from source" build without actually building GCC and friends.

3) `USE_SYSTEM_CSL=1`: Symlinks the libraries into `usr/lib/julia` using the same logic as the rest of the `USE_SYSTEM_*` flags.

I think this should satisfy all use cases.  :)

yuyichao commented:
The revised behavior SGTM.

staticfloat committed:
Expand `JL_PRIVATE_LIBS-1` for versioned library names as well

Without this, we fail to install things like `libgcc_s.so.1` when
installed into the private libdir by `symlink_system_libraries`.

ronisbr commented:
I am not sure if this is related, but using this patch in 1.6 and building Julia with `USE_SYSTEM_CSL=1`, the system tries to build the documentation even if it is already present from the tarball I downloaded:

```
[ 2343s] /home/abuild/rpmbuild/BUILD/julia-1.6.0/deps/tools/jldownload "/home/abuild/rpmbuild/BUILD/julia-1.6.0/deps/srccache/UnicodeData-13.0.0.txt" http://www.unicode.org/Public/13.0.0/ucd/UnicodeData.txt
[ 2343s] Could not find working curl, wget, or fetch.
[ 2343s] You need to install one of these to download dependencies.
[ 2343s] make[2]: *** [Makefile:30: /home/abuild/rpmbuild/BUILD/julia-1.6.0/deps/srccache/UnicodeData-13.0.0.txt] Error 1
[ 2343s] make[2]: Leaving directory '/home/abuild/rpmbuild/BUILD/julia-1.6.0/doc'
[ 2343s] make[1]: *** [Makefile:95: docs] Error 2
[ 2343s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/julia-1.6.0'
[ 2343s] make: *** [Makefile:47: /home/abuild/rpmbuild/BUILD/julia-1.6.0/doc/_build/html/en/index.html] Error 2
[ 2343s] error: Bad exit status from /var/tmp/rpm-tmp.R86M1y (%install)
```

The build works fine without this patch and `USE_SYSTEM_CSL=0`.
